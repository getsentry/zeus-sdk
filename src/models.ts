/** Generic status for builds, jobs or artifacts. */
export enum Status {
  /** No status is known. */
  UNKNOWN = 'unknown',
  /** The entity is queued for execution or processing. */
  QUEUED = 'queued',
  /** The entity is currently being processed. */
  PROGRESS = 'in_progress',
  /** Execution has finished and processing is complete. */
  FINISHED = 'finished',
  /** Execution has finished, but results are still being collected. */
  COLLECTING = 'collecting_results',
  /** Only for artifacts: The artifact is no longer available for download. */
  EXPIRED = 'expired',
}

/** Result of a finished job or build. */
export enum Result {
  /** No result is known. */
  UNKNOWN = 'unknown',
  /** The job or build was aborted. */
  ABORTED = 'aborted',
  /** The job or build has passed successfully. */
  PASSED = 'passed',
  /** The job or build did not execute. */
  SKIPPED = 'skipped',
  /** The job or build exited with an unexpected error. */
  ERRORED = 'errored',
  /** The job or build failed (likely in tests). */
  FAILED = 'failed',
}

/** A build artifact uploaded to Zeus. */
export interface Artifact {
  /** Unique identifier of the artifact in Zeus. */
  id: string;
  /** File name used for downloading the file. */
  name: string;
  /** File mime type. Does not use valid IETF RFC 6838 types currently. */
  type: string;
  /** Fully qualified download URL for the artifact file. */
  download_url: string;
  /** Processing status of this artifact. */
  status: Status;
  /** ISO timestamp of the creation date. */
  created_at?: string;
  /** ISO timestamp of the last update. */
  updated_at?: string;
  /** ISO timestamp when processing has started. */
  started_at?: string;
  /** ISO timestamp when processing has finished. */
  finished_at?: string;
}

/** The author of a source or revision. */
export interface Author {
  /** Unique author identifier. */
  id: string;
  /** Display name of the author. */
  name: string;
  /** Optional email address. */
  email?: string;
}

/** Source revision that is the base for a build. */
export interface Revision {
  /** Unique revision identifier. */
  id: string;
  /** VCS revision identifier, usually a unique number or hash. */
  sha: string;
  /** The author's message attached to this revision. */
  message: string;
  /** Author of this revision. */
  author: Author;
  /** ISO timestamp when this revision was added to Zeus. */
  created_at: string;
  /** ISO timestamp when this revision was created by the author. */
  committed_at?: string;
}

/** A commit or patch that triggers a build. */
export interface Source {
  /** Unique source identifier. */
  id: string;
  /** The source revision (e.g. VCS commit). */
  revision: Revision;
  /** The author of this source. */
  author: Author;
  /** ISO timestamp when this source was added to Zeus. */
  created_at: string;
}

/** Aggregated revision information as returned by revisions API */
export interface RevisionInfo {
  /** Unique revision identifier. */
  id: string;
  /** Source information for the given revision. */
  source: Source;
  /** Revision status. */
  status: Status;
  /** Revision result. */
  result: Result;
  /** Human readable display label of the revision. */
  label: string;
  /** ISO timestamp when this source was added to Zeus. */
  created_at: string;
  /** ISO timestamp when the first build for the revision started. */
  started_at?: string;
  /** ISO timestamp when the last build for the revision finished. */
  finished_at?: string;
  /** Aggregate statistics of revision builds. */
  stats: Stats;
}

/** Code coverage statistics. */
export interface CoverageStats {
  /** The total number of source lines covered by tests. */
  lines_covered: number;
  /** The total number of source lines not covered by tests. */
  lines_uncovered: number;
  /** The number of changed source lines covered by tests. */
  diff_lines_covered: number;
  /** The number of changed source lines not covered by tests. */
  diff_lines_uncovered: number;
}

/** Test statistics. */
export interface TestStats {
  /** Total number of tests executed. */
  count: number;
  /** Total number of failed tests. */
  failures: number;
  /** Test duration in seconds. */
  duration: number;
  /** Deduplicated number of tests executed. */
  count_unique: number;
  /** Deduplicated number of failed tests. */
  failures_unique: number;
}

/** Source code style statistics. */
export interface StyleStats {
  /** Number of style violations in the code base. */
  count: number;
}

/** Webpack asset statistics. */
export interface WebpackStats {
  /** The aggregate size of all assets in bytes. */
  total_asset_size: number;
}

/** Build or Job statistics computed by processing artifacts. */
export interface Stats {
  coverage: CoverageStats;
  tests: TestStats;
  style_violations: StyleStats;
  webpack: WebpackStats;
}

/** A build on CI. */
export interface Build {
  /** Unique build identifier. */
  id: string;
  /** Build number assigned by Zeus, unique to this project. */
  number: number;
  /** Human readable display label of the build. */
  label: string;
  /** Identifier of the CI provider. */
  provider: string;
  /** The build identifier in the provider CI system. */
  external_id: string;
  /** Fully qualified URL to the build on the provider. */
  url: string;
  /** Build status. */
  status: Status;
  /** Build result. */
  result: Result;
  /** Reference to the code that has been built (VCS commit or patch). */
  source?: Source;
  /** Aggregate statistics of build results. */
  stats?: Stats;
  /** ISO timestamp when the build was created (queued). */
  created_at: string;
  /** ISO timestamp when the first job of the build started. */
  started_at?: string;
  /** ISO timestamp when all jobs of the build have finished. */
  finished_at?: string;
}

/** The status of a job or build on Zeus. */
export enum JobStatus {
  /** The job or build is still running. */
  PENDING = 'pending',
  /** The job or build has succeeded. */
  PASSED = 'passed',
  /** The job or build has failed. */
  FAILED = 'failed',
}

/** Details to a failed job. */
export interface Failure {
  /** Human readable failure description. */
  reason: string;
}

/** A job on CI. */
export interface Job {
  /** Unique job identifier */
  id: string;
  /** Job number assigned by Zeus, scoped to its build. */
  number: number;
  /** Human readable display label of the job. */
  label: string;
  /** Identifier of the CI provider. */
  provider: string;
  /** The job identifier in the provider CI system. */
  external_id: string;
  /** Fully qualified URL to the job on the provider. */
  url: string;
  /** Whether this job is allowed failure. */
  allow_failure: boolean;
  /** Job status. */
  status: Status;
  /** Job result. */
  result: Result;
  /** Aggregate statistics of job results. */
  stats?: Stats;
  /** Failure details if this job failed. */
  failures: Failure[];
  /** ISO timestamp when the job was created (queued). */
  created_at: string;
  /** ISO timestamp when the job was last updated created. */
  updated_at?: string;
  /** ISO timestamp when the job started executing. */
  started_at?: string;
  /** ISO timestamp when job has finished executing. */
  finished_at?: string;
}
