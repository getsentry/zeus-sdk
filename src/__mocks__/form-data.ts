'use strict';

module.exports = class FormData {
  private readonly data: Record<string, any>;

  public constructor() {
    this.data = {};
  }

  public append(field: string, value: any, options: any): void {
    this.data[field] = { value, options };
  }

  public getHeaders(): Record<string, string> {
    return {
      'content-type': 'multipart/form-data; boundary=---feedface',
    };
  }

  public getLength(cb: (e?: Error, len?: number) => void): void {
    cb(undefined, 42);
  }
};
