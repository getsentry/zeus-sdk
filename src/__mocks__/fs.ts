function createWriteStream(): any {
  return {
    on: (action, callback) => {
      if (action === 'finish') {
        callback();
      }
    },
  };
}

export { createWriteStream };
