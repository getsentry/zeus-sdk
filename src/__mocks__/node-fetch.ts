// tslint:disable-next-line:no-default-export
export default jest.fn();
