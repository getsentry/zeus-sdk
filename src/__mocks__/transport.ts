export class Transport {
  public static instance: Transport;
  public postForm: jest.Mock;
  public request: jest.Mock;
  public requestJson: jest.Mock;

  public constructor() {
    Transport.instance = this;
    this.postForm = jest.fn();
    this.request = jest.fn();
    this.requestJson = jest.fn();
  }
}
