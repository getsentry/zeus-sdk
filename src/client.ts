import { join } from 'path';
import { URL } from 'whatwg-url';
import { Artifact, Build, Job, JobStatus, Result, Status } from './models';
import { FormField, Options, Transport } from './transport';
import { getHookBase, sanitizeURL } from './utils';

/** Default URL of the Zeus server. */
export const DEFAULT_URL = 'https://zeus.ci/';

/** Options passed to `Client.uploadArtifact`. */
export interface UploadOptions {
  /** The build identifier. */
  build: string;
  /** The job identifier. */
  job: string;
  /** An artifact file to upload. */
  file: string | Buffer;
  /** Optional MIME type of the artifact file. */
  type?: string;
  /** Optional name of the artifcat file. */
  name?: string;
}

/** Options passed to `Client.createOrUpdateBuild`. */
export interface BuildOptions {
  /** The build identifier. */
  build: string;
  /** Display label for this build. If empty, Zeus will infer one. */
  buildLabel?: string;
  /** Fully qualified URL to the build on the CI system. */
  url?: string;
  /** Reference to the VCS commit triggering the build. */
  ref?: string;
}

/** Options passed to `Client.createOrUpdateJob`. */
export interface JobOptions {
  /** The job identifier. */
  job: string;
  /** The build identifier. */
  build: string;
  /** Display label for this job. If empty, Zeus will infer one. */
  jobLabel?: string;
  /** Job resolution status. */
  status?: JobStatus;
  /** Fully qualified URL to the job on the CI system. */
  url?: string;
  /** Reference to the VCS commit triggering the build. */
  ref?: string;
}

/**
 * Zeus API client.
 */
export class Client {
  /** Internal transport used to send requests to Zeus. */
  private readonly transport: Transport;
  public readonly owner: string;
  public readonly repo: string;

  /**
   * Creates a new API client.
   *
   * @param options Optional configuration parameters for the client.
   */
  public constructor(options: Options = {}) {
    this.transport = new Transport(options);
    this.owner = options.owner || '';
    this.repo = options.repo || '';
  }

  /**
   * Uploads one or more build artifacts for a given build job.
   *
   * @param options Parameters for the upload request.
   * @returns The parsed server response.
   */
  public async uploadArtifact(options: UploadOptions): Promise<Artifact> {
    const base = getHookBase();
    if (!options) {
      throw new Error('Missing upload options');
    } else if (!options.build) {
      throw new Error('Missing build identifier');
    } else if (!options.job) {
      throw new Error('Missing job identifier');
    } else if (!options.file) {
      throw new Error('Missing file parameter');
    }

    const url = new URL(
      `builds/${options.build}/jobs/${options.job}/artifacts`,
      sanitizeURL(base)
    ).toString();

    return this.transport.postForm<Artifact>(url, {
      file: [options.file, options.name] as FormField,
      type: options.type,
    });
  }

  /**
   * Creates or updates the remote build object.
   *
   * @param options Parameters for the build.
   * @returns The parsed server response.
   */
  public async createOrUpdateBuild(options: BuildOptions): Promise<Build> {
    const base = getHookBase();
    if (!options.build) {
      throw new Error('Missing build ID');
    }

    const url = new URL(
      `builds/${options.build}`,
      sanitizeURL(base)
    ).toString();

    return this.transport.requestJson<Build>(url, {
      body: {
        label: options.buildLabel,
        ref: options.ref,
        url: options.url,
      },
      method: 'POST',
    });
  }

  /**
   * Creates or updates the remote job object.
   *
   * @param options Parameters for the job.
   * @returns The parsed server response.
   */
  public async createOrUpdateJob(options: JobOptions): Promise<Job> {
    const base = getHookBase();
    if (!options.job) {
      throw new Error('Missing job ID');
    }

    // Deal with statuses. Zeus has knowledge about both 'result' and
    // 'status' for every job, but to make it easier for end user we'll
    // allow to specify only a handful of unified 'statuses'.
    let status: string | undefined;
    let result: string | undefined;

    switch (options.status) {
      case JobStatus.PENDING:
        status = Status.PROGRESS;
        result = Result.UNKNOWN;
        break;
      case JobStatus.PASSED:
        status = Status.FINISHED;
        result = Result.PASSED;
        break;
      case JobStatus.FAILED:
        status = Status.FINISHED;
        result = Result.FAILED;
        break;
      case undefined:
        break;
      default:
        throw new Error(`Invalid job status: "${options.status}"`);
    }

    const url = new URL(
      `builds/${options.build}/jobs/${options.job}`,
      sanitizeURL(base)
    ).toString();

    return this.transport.requestJson<Job>(url, {
      body: {
        label: options.jobLabel,
        ref: options.ref,
        result,
        status,
        url: options.url,
      },
      method: 'POST',
    });
  }

  /**
   * Computes the absolute URL of an endpoint specified by the given path.
   *
   * @param path Path to an endpoint.
   * @returns The absolute URL to the endpoint.
   */
  public getUrl(path: string = ''): string {
    return this.transport.getUrl(path);
  }

  ///////////////////////////////////////////////

  /**
   * Downloads all files stored for the commit
   *
   * Retrieves the full list of artifacts from Zeus and stores them in the
   * download directory. Each file is only downloaded once when invoked
   * multiple times.
   *
   * @returns Absolute paths to local copies of all files
   * @async
   */
  // public async downloadAll() {
  //   const files = await listFiles();
  //   return downloadFiles(files);
  // }

  /**
   * Downloads a file from the store
   *
   * The file is placed in the download directory. It is only downloaded once
   * when invoked multiple times. If the file does not exist, an error is
   * thrown. Use {@link listFiles} to retrieve available files.
   *
   * @param file A file object to download
   * @returns Absolute path to the local copy of the file
   * @async
   */
  public async downloadArtifact(file: Artifact): Promise<any> {
    const url = this.getUrl(file.download_url);
    console.log('url', url);
    // logger.debug(`Downloading Zeus file ${url} to ${downloadDirectory}`);

    const localFile = join('/tmp/zeus-downloads', file.name);
    console.log('local file', localFile);

    await fetch(url).then(response => console.log(response));

    return new Promise((resolve, reject) => {
      resolve('');
      reject('');
    });
    // const stream = await this.transport
    //   .request(url)
    //   .pipe(createWriteStream(localFile));

    // const promise = new Promise((resolve, reject) => {
    //   // NOTE: The timeout is necessary to be able to list files immediately
    //   stream.on('finish', () => setTimeout(() => resolve(localFile), 100));
    //   stream.on('error', reject);
    // });

    // return promise;
  }

  /**
   * Retrieves a list of files stored for the commit
   *
   * The list is only loaded once if invoked multiple times.
   *
   * @returns A list of file objects
   * @async
   */
  public async listFilesForRevision(sha: string): Promise<Artifact[]> {
    const url = `/api/repos/gh/${this.owner}/${
      this.repo
    }/revisions/${sha}/artifacts`;
    return this.transport.request<Artifact[]>(url);
  }

  /**
   * Retrieves a list of files stored for the build
   *
   * The list is only loaded once if invoked multiple times.
   *
   * @returns A list of file objects
   * @async
   */
  public async listFilesForBuild(buildNumber: number): Promise<Artifact[]> {
    const url = `/api/repos/gh/${this.owner}/${
      this.repo
    }/builds/${buildNumber}/artifacts`;
    return this.transport.request<Artifact[]>(url);
  }

  /**
   * Retrieves a list of files stored for the job
   *
   * The list is only loaded once if invoked multiple times.
   *
   * @returns A list of file objects
   * @async
   */
  public async listFilesForJob(
    buildNumber: number,
    jobNumber: number
  ): Promise<Artifact[]> {
    const url = `/api/repos/gh/${this.owner}/${
      this.repo
    }/builds/${buildNumber}/jobs/${jobNumber}/artifacts`;
    return this.transport.request<Artifact[]>(url);
  }
}
