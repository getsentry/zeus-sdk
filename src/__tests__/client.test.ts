import { existsSync, lstatSync } from 'fs';
import { Transport } from '../__mocks__/transport';
import { Artifact, JobStatus, Status } from '../models';

// Mock the transport module and make sure we've imported the same instance.
// The name must start with "mock", to bypass jest's module factory protection.
const mockTransport = Transport;
jest.mock('../transport', () => ({ Transport: mockTransport }));
jest.mock('fs');

import { BuildOptions, Client, JobOptions, UploadOptions } from '../client';

const existsSyncMock = existsSync as jest.Mock<typeof existsSync>;
const lstatSyncMock = lstatSync as jest.Mock<typeof lstatSync>;

describe('Client', () => {
  const env = { ...process.env };
  let client: Client;

  beforeEach(() => {
    process.env.ZEUS_HOOK_BASE = 'https://example.org/hooks/feedbeef/';
    client = new Client();
  });

  afterEach(() => {
    process.env = { ...env };
  });

  describe('uploadArtifact', () => {
    let params: UploadOptions;

    beforeEach(() => {
      params = {
        build: '12345',
        file: 'FILE_DATA',
        job: '54321',
      };
    });

    test('uploads a file without explicit type', async () => {
      await client.uploadArtifact(params);
      expect(mockTransport.instance.postForm).toMatchSnapshot();
    });

    test('uploads a file with explicit type', async () => {
      params.type = 'text/plain';
      await client.uploadArtifact(params);
      expect(mockTransport.instance.postForm).toMatchSnapshot();
    });

    test('uploads a file with explicit name', async () => {
      params.name = 'renamed.json';
      await client.uploadArtifact(params);
      expect(mockTransport.instance.postForm).toMatchSnapshot();
    });

    test('accepts ZEUS_HOOK_BASE without trailing slash', async () => {
      process.env.ZEUS_HOOK_BASE = 'https://example.org/hooks/feedbeef';
      await client.uploadArtifact(params);
      expect(mockTransport.instance.postForm).toMatchSnapshot();
    });

    test('resolves the server response', async () => {
      const data = { id: 4711, download_url: '/foo/bar' };
      mockTransport.instance.postForm.mockReturnValue(Promise.resolve(data));
      const response = await client.uploadArtifact(params);
      expect(response).toEqual(data);
    });

    test('rejects without parameters', async () => {
      try {
        await client.uploadArtifact(undefined as any);
      } catch (e) {
        // This should probably become an error message at some point
        expect((e as Error).toString()).toMatch(/options/i);
      }
    });

    test('rejects without ZEUS_HOOK_BASE', async () => {
      delete process.env.ZEUS_HOOK_BASE;
      try {
        await client.uploadArtifact(params);
      } catch (e) {
        expect((e as Error).toString()).toMatch('ZEUS_HOOK_BASE');
      }
    });

    test('rejects an invalid ZEUS_HOOK_BASE', async () => {
      process.env.ZEUS_HOOK_BASE = '///';
      try {
        await client.uploadArtifact(params);
      } catch (e) {
        expect((e as Error).toString()).toMatch(/invalid url/i);
      }
    });

    test('rejects without the build parameter', async () => {
      delete params.build;
      try {
        await client.uploadArtifact(params);
      } catch (e) {
        expect((e as Error).toString()).toMatch(/build/i);
      }
    });

    test('rejects without the job parameter', async () => {
      delete params.job;
      try {
        await client.uploadArtifact(params);
      } catch (e) {
        expect((e as Error).toString()).toMatch(/job/i);
      }
    });

    test('rejects without the file parameter', async () => {
      delete params.file;
      try {
        await client.uploadArtifact(params);
      } catch (e) {
        expect((e as Error).toString()).toMatch(/file/i);
      }
    });
  });

  describe('createOrUpdateBuild', () => {
    let params: BuildOptions;

    beforeEach(() => {
      params = {
        build: '1234',
        url: 'https://ci/build/1234',
      };
    });

    test('resolves the server response', async () => {
      const data = { id: '1234' };
      mockTransport.instance.requestJson.mockReturnValue(data);
      const response = await client.createOrUpdateBuild(params);
      expect(response).toEqual(data);
    });

    test('rejects without the build parameter', async () => {
      delete params.build;
      try {
        await client.createOrUpdateBuild(params);
      } catch (e) {
        expect((e as Error).toString()).toMatch(/build id/i);
      }
    });

    test('rejects without ZEUS_HOOK_BASE', async () => {
      delete process.env.ZEUS_HOOK_BASE;
      try {
        await client.createOrUpdateBuild(params);
      } catch (e) {
        expect((e as Error).toString()).toMatch('ZEUS_HOOK_BASE');
      }
    });

    test('passes build label properly', async () => {
      params.buildLabel = 'new build label';
      await client.createOrUpdateBuild(params);
      expect(mockTransport.instance.requestJson).toMatchSnapshot();
    });

    test('strips invalid build attributes', async () => {
      const options = { ...params, invalidAttribute: 'test' };
      await client.createOrUpdateBuild(options);
      expect(mockTransport.instance.requestJson).toMatchSnapshot();
    });
  });

  describe('createOrUpdateJob', () => {
    let params: JobOptions;

    beforeEach(() => {
      params = {
        build: '1234',
        job: '2345',
        url: 'https://ci/build/1234',
      };
    });

    test('resolves the server response', async () => {
      const data = { id: '1234' };
      mockTransport.instance.requestJson.mockReturnValue(data);
      const response = await client.createOrUpdateJob(params);
      expect(response).toEqual(data);
    });

    test('rejects without ZEUS_HOOK_BASE', async () => {
      delete process.env.ZEUS_HOOK_BASE;
      try {
        await client.createOrUpdateJob(params);
      } catch (e) {
        expect((e as Error).toString()).toMatch('ZEUS_HOOK_BASE');
      }
    });

    test('rejects an invalid ZEUS_HOOK_BASE', async () => {
      process.env.ZEUS_HOOK_BASE = '///';
      try {
        await client.createOrUpdateJob(params);
      } catch (e) {
        expect((e as Error).toString()).toMatch(/invalid url/i);
      }
    });

    test('rejects without the job parameter', async () => {
      delete params.job;
      try {
        await client.createOrUpdateJob(params);
      } catch (e) {
        expect((e as Error).toString()).toMatch(/job id/i);
      }
    });

    test('rejects with invalid job status', async () => {
      params.status = 'invalid-status' as JobStatus;
      try {
        await client.createOrUpdateJob(params);
      } catch (e) {
        expect((e as Error).toString()).toMatch(params.status);
      }
    });

    test('sets the "pending" status', async () => {
      params.status = JobStatus.PENDING;
      await client.createOrUpdateJob(params);
      expect(mockTransport.instance.requestJson).toMatchSnapshot();
    });

    test('sets the "passed" status', async () => {
      params.status = JobStatus.PASSED;
      await client.createOrUpdateJob(params);
      expect(mockTransport.instance.requestJson).toMatchSnapshot();
    });

    test('sets the "failed" status', async () => {
      params.status = JobStatus.FAILED;
      await client.createOrUpdateJob(params);
      expect(mockTransport.instance.requestJson).toMatchSnapshot();
    });

    test('passes job label properly', async () => {
      params.jobLabel = 'new job label';
      await client.createOrUpdateJob(params);
      expect(mockTransport.instance.requestJson).toMatchSnapshot();
    });

    test('strips invalid job attributes', async () => {
      (params as any).invalidAttribute = 'test';
      await client.createOrUpdateJob(params);
      expect(mockTransport.instance.requestJson).toMatchSnapshot();
    });
  });

  describe('getUrl', () => {
    test('calls getUrl from transport', async () => {
      const path = '/build/test';
      client.getUrl(path);
      expect(mockTransport.instance.getUrl).toBeCalledWith(path);
    });
  });

  describe('listFiles', () => {
    const owner = 'getsentry';
    const repo = 'craft';
    beforeEach(() => {
      client = new Client({ owner, repo });
    });

    test('listFilesForRevision calls correct URL', async () => {
      const sha = '8c99609eeed9dc3297728411b9306e2c3786e127';
      const expectedUrl = `/api/repos/gh/getsentry/craft/revisions/${sha}/artifacts`;
      await client.listArtifactsForRevision(owner, repo, sha);
      expect(mockTransport.instance.request).toBeCalledWith(expectedUrl);
    });

    test('listFilesForBuild calls correct URL', async () => {
      const buildNumber = 123;
      const expectedUrl = `/api/repos/gh/getsentry/craft/builds/${buildNumber}/artifacts`;
      await client.listArtifactsForBuild(owner, repo, buildNumber);
      expect(mockTransport.instance.request).toBeCalledWith(expectedUrl);
    });

    test('listFilesForJob calls correct URL', async () => {
      const buildNumber = 123;
      const jobNumber = 234;
      const expectedUrl = `/api/repos/gh/getsentry/craft/builds/${buildNumber}/jobs/${jobNumber}/artifacts`;
      await client.listArtifactsForJob(owner, repo, buildNumber, jobNumber);
      expect(mockTransport.instance.request).toBeCalledWith(expectedUrl);
    });
  });

  describe('downloadArtifact', () => {
    const defaultDirectory = '/downloads';
    const fileName = 'file.txt';
    const artifact: Artifact = {
      download_url: 'http://zeus/download/file.txt',
      id: '123',
      name: fileName,
      status: Status.UNKNOWN,
      type: 'text/plain',
    };
    beforeEach(() => {
      client = new Client({ defaultDirectory });
    });

    test('saves artifact to a file', async () => {
      expect.assertions(1);
      mockTransport.instance.requestRaw.mockImplementation(() => ({
        body: {
          on: jest.fn(),
          pipe: jest.fn(),
        },
      }));

      // tslint:disable-next-line
      existsSyncMock.mockReturnValue(true);
      lstatSyncMock.mockReturnValue({ isDirectory: () => true });

      const filePath = await client.downloadArtifact(artifact);

      expect(filePath).toBe(`${defaultDirectory}/${fileName}`);
    });

    test('returns error if downloadDirectory is not provided', async () => {
      expect.assertions(1);
      client = new Client();
      expect(client.downloadArtifact(artifact)).rejects.toThrowError(
        /directory not specified/i
      );
    });
  });
});
