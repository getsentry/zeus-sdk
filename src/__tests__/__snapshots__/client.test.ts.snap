// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Client createOrUpdateBuild passes build label properly 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "https://example.org/hooks/feedbeef/builds/1234",
      Object {
        "body": Object {
          "label": "new build label",
          "ref": undefined,
          "url": "https://ci/build/1234",
        },
        "method": "POST",
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;

exports[`Client createOrUpdateBuild strips invalid build attributes 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "https://example.org/hooks/feedbeef/builds/1234",
      Object {
        "body": Object {
          "label": undefined,
          "ref": undefined,
          "url": "https://ci/build/1234",
        },
        "method": "POST",
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;

exports[`Client createOrUpdateJob passes job label properly 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "https://example.org/hooks/feedbeef/builds/1234/jobs/2345",
      Object {
        "body": Object {
          "label": "new job label",
          "result": undefined,
          "status": undefined,
          "url": "https://ci/build/1234",
        },
        "method": "POST",
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;

exports[`Client createOrUpdateJob sets the "failed" status 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "https://example.org/hooks/feedbeef/builds/1234/jobs/2345",
      Object {
        "body": Object {
          "label": undefined,
          "result": "failed",
          "status": "finished",
          "url": "https://ci/build/1234",
        },
        "method": "POST",
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;

exports[`Client createOrUpdateJob sets the "passed" status 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "https://example.org/hooks/feedbeef/builds/1234/jobs/2345",
      Object {
        "body": Object {
          "label": undefined,
          "result": "passed",
          "status": "finished",
          "url": "https://ci/build/1234",
        },
        "method": "POST",
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;

exports[`Client createOrUpdateJob sets the "pending" status 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "https://example.org/hooks/feedbeef/builds/1234/jobs/2345",
      Object {
        "body": Object {
          "label": undefined,
          "result": "unknown",
          "status": "in_progress",
          "url": "https://ci/build/1234",
        },
        "method": "POST",
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;

exports[`Client createOrUpdateJob strips invalid job attributes 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "https://example.org/hooks/feedbeef/builds/1234/jobs/2345",
      Object {
        "body": Object {
          "label": undefined,
          "result": undefined,
          "status": undefined,
          "url": "https://ci/build/1234",
        },
        "method": "POST",
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;

exports[`Client uploadArtifact accepts ZEUS_HOOK_BASE without trailing slash 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "https://example.org/hooks/feedbeef/builds/12345/jobs/54321/artifacts",
      Object {
        "file": Array [
          "FILE_DATA",
          undefined,
        ],
        "type": undefined,
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;

exports[`Client uploadArtifact uploads a file with explicit name 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "https://example.org/hooks/feedbeef/builds/12345/jobs/54321/artifacts",
      Object {
        "file": Array [
          "FILE_DATA",
          "renamed.json",
        ],
        "type": undefined,
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;

exports[`Client uploadArtifact uploads a file with explicit type 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "https://example.org/hooks/feedbeef/builds/12345/jobs/54321/artifacts",
      Object {
        "file": Array [
          "FILE_DATA",
          undefined,
        ],
        "type": "text/plain",
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;

exports[`Client uploadArtifact uploads a file without explicit type 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "https://example.org/hooks/feedbeef/builds/12345/jobs/54321/artifacts",
      Object {
        "file": Array [
          "FILE_DATA",
          undefined,
        ],
        "type": undefined,
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;
